src/rdkafka.c

```diff
@@ -3711,6 +3711,9 @@ int rd_kafka_poll (rd_kafka_t *rk, int timeout_ms) {
         return r;
 }
 
+void rd_kafka_poll_wake (rd_kafka_t *rk) {
+        rd_kafka_q_wake(rk->rk_rep);
```

Comment:
There's already rd_kafka_q_yield() that does the same thing, could use that here instead.
Might be a good idea to change the cnd_signal() to cnd_broadcast() in q_yield().

Reply:
Bearing in mind that I have very little idea how any of this works ;) do you think it might make sense to provide _both_ 
- `rd_kafka_q_wake_one()`, formerly known as `rd_kafka_q_yield()`
- `rd_kafka_q_wake_all()`, formerly known as `rd_kafka_q_wake()`

so that we can use the latter here, without accidentally breaking or pessimizing any of the callers of the former?

Reply:
I think q_yield() should be using a broadcast instead of a signal, making q_yield and q_wake identical.

Reply:
Well, I'm still attached to my idea, but as I don't actually know this code at all, I defer to your expertise. :) Do you want me to amend this patch along those lines, or are you planning to change the behavior of `rd_kafka_q_yield` in master regardless, or what's the deal at this point?

Reply:
Please go ahead and change q_yield() to use broadcast instead of signal.

Reply:
Updated. I should perhaps have pointed out before updating that you said "making q_yield and q_wake identical" but that doesn't seem right at all; their code is VERY different.

Reply:
Funcionality wise they are identical, so I don't see the need for a q_wake() or EXPLICIT_WAKE, that's already what q_yield() does.

Reply:
Hmm, then I don't understand what `rd_kafka_q_yield()` does. The intent behind this PR #3114 is that we have a caller roughly like this:

```
class Producer {
    std::thread t_;
    std::unique_ptr<rd_kafka_t, RdKafkaDeleter> rk_;
    std::atomic<bool> done_;
public:
    explicit Producer() : done_(false) {
        rk_ = ...snip...;
        t_ = std::thread([]() {
            while (!done_) {
                rd_kafka_poll(rk_.get(), -1);
            }
        });
    }
    void shutdown() {
        if (!done_.exchange(true)) {
            rd_kafka_poll_wake();  // Force rd_kafka_poll() to return to its caller
            t_.join();
        }
    }
    ~Producer() { shutdown(); }
};
```
My impression from looking at your code was that `rd_kafka_q_yield()` doesn't actually cause `rd_kafka_poll()` to return to its caller unless there's work to do; that's why we had to introduce the new `RD_KAFKA_Q_F_EXPLICIT_WAKE` flag, to say "If you woke up and there's no work to do, _still_ I need you to treat this particular wakeup as non-spurious and return to your caller, because the program explicitly requested this wakeup."

- What should we do in `Producer::shutdown()` instead of `rd_kafka_poll_wake()`? what one-liner should go there instead?

- Or is `class Producer` itself wrong-headed? Should it be _completely_ differently architected? and if so, how?

Reply:
Have you tried using q_yield()?

Reply:
(Notice that I had at least one typo above: there should be an argument to `rd_kafka_poll_wake(...)`!)

Are you suggesting that the `class Producer` above would work fine if I replace the line `rd_kafka_poll_wake();` with `rd_kafka_q_yield(rk_.get());`? Is this how rdkafka is intended to be used?